components/tidb_query_expr/src/impl_like.rs

```diff
@@ -35,13 +39,13 @@ pub fn like<C: Collator>(target: BytesRef, pattern: BytesRef, escape: &i64) -> R
             } else {
                 if code == escape && px + poff < pattern.len() {
                     px += poff;
-                    poff = if let Some((_, off)) = C::Charset::decode_one(&pattern[px..]) {
+                    poff = if let Some((_, off)) = PatternCharset::decode_one(&pattern[px..]) {
                         off
                     } else {
                         break;
                     }
                 }
-                if let Some((_, toff)) = C::Charset::decode_one(&target[tx..]) {
+                if let Some((_, toff)) = TargetCharset::decode_one(&target[tx..]) {
                     if let Ok(std::cmp::Ordering::Equal) =
                         C::sort_compare(&target[tx..tx + toff], &pattern[px..px + poff])
```

Comment:
If pattern and target have different charsets, does this comparison still make sense :thinking: ?

Reply:
It's technically fine for TiKV in this situation, but not actually correct...

TiKV only have two charset format (for storage): binary or utf8mb4, and we only need to know whether they are equal, then using any specified collator to compare them is fine :thinking:.

If TiKV supports more charset in the storage (e.g. GBK, ...), this function need to be reworked (anyway, the original implementation also needs, because you cannot use a single charset to decode two binaries with different charsets).

Reply:
I'll add some comments to describe the limitation.